Orange/projection/freeviz.py

```diff
@@ -0,0 +1,467 @@
+
+    def compute(self, data, freeviz_space):
+        if self.feature >= freeviz_space.shape[1]:
+            return np.zeros((freeviz_space.shape[0]))
+        return freeviz_space[:, self.feature]
+
+
+class _FreeVizTransformDomain:
+    """Computation common for all FreeViz variables."""
+
+    def __init__(self, freeviz):
+        self.freeviz = freeviz
+
+    def __call__(self, data):
+        if data.domain != self.freeviz.pre_domain:
+            data = data.transform(self.freeviz.pre_domain)
+        return self.freeviz.transform(data.X)
+
+
+class _LinearCombination:
+    def __init__(self, attrs, weights, mean=None):
+        self.attrs = attrs
+        self.weights = weights
+        self.mean = mean
+
+
+class FreeVizModel(Projection):
+    name = "FreeVizModel"
+
+    def __init__(self, proj, domain):
+        freeviz_transform = _FreeVizTransformDomain(self)
+
+        def freeviz_variable(i):
+            v = ContinuousVariable(
+                "FreeViz Component {}".format(i + 1),
+                compute_value=FreeVizProjector(self, i, freeviz_transform))
+            v.to_sql = _LinearCombination(
+                domain.attributes, self.components_[i, :],
+                getattr(self, 'mean_', None))
+            return v
+
+        super().__init__(proj=proj)
+        self.orig_domain = domain
+        self.n_components = self.components_.shape[0]
+        self.domain = Domain(
+            [freeviz_variable(i) for i in range(self.n_components)],
+            domain.class_vars, domain.metas)
+
+
+class FreeViz(Projector):
+    name = 'FreeViz'
+    supports_sparse = False
+
+    def __init__(self, weights=None, center=True, scale=True, dim=2, p=1,
+                 initial=None, maxiter=500, alpha=0.1, atol=1e-5, preprocessors=None):
+        super().__init__(preprocessors=preprocessors)
+        self.params = vars()
+        self.weights = weights
+        self.center = center
+        self.scale = scale
+        self.dim = dim
+        self.p = p
+        self.initial = initial
+        self.maxiter = maxiter
+        self.alpha = alpha
+        self.atol = atol
+        self.is_class_discrete = False
+
+    def __call__(self, data):
+        if data and data.domain.class_var.is_discrete:
+            self.is_class_discrete = True
+        return super().__call__(data)
+
+    def fit(self, X, Y=None):
+        X, Y, _ = self.prepare_freeviz_data(X=X, Y=Y)
+        if X is not None:
+            _, a, _, _ = self.freeviz(X, Y, weights=self.weights, center=self.center,
+                                      scale=self.scale, dim=self.dim, p=self.p,
+                                      initial=self.initial, maxiter=self.maxiter, alpha=self.alpha,
+                                      atol=self.atol, is_class_discrete=self.is_class_discrete)
+            self.components_ = a
+            return FreeVizModel(self, self.domain)
+
+    def transform(self, X):
+        EX = np.dot(X, self.components_)
+        return EX
+
```

Comment:
@markotoplak, the code from the beginning of the file and up to here is adapted from PCA. I know the idea, but you're an expert here. Can you take a quick look, please?

Reply:
Seems OK.

I think it would be useful to add a test, where we would
1. open data, apply some preprocessor, splits the data into two parts, use FreeViz on the first part, and then transform the second part,
2. open data, split into two parts, apply the same preprocessor and FreeViz only on the first part, and then transform the second part,

where the transformed second part in (1) and (2) has to be the same.